### ISSUE TITLE
Potential memory/goroutines leaks

### ISSUE THREAD

**Issue by gaelL** [concourse/concourse#8638]
## Summary

We upgraded from 6.7..6 to 7.8.3.
Our concourse web have a bunch of pipelines, teams and dedicated team workers.

Since the upgrade we identified memory growing until it got killed by OOM 

And in prometheus metrics we can see goroutines number increase too.

![image](https://user-images.githubusercontent.com/2479183/205684076-540880ad-c1c8-435b-a210-9d01ea381327.png)

![image](https://user-images.githubusercontent.com/2479183/205684089-9e4d18f9-a318-4767-bc6f-f0f549f682ea.png)

And something interesting

![image](https://user-images.githubusercontent.com/2479183/205686354-618c8319-0d7c-406f-9c6a-497cbffaf348.png)

## Steps to reproduce

Not sure

## Additional context

This Concourse server run only one web component. This instance is running since Concourse version 4 and have quite lot of team and pipelines.
On this Concourse several workers are attached to specific teams (there is no global/default workers).
The database is around 8Go, we have quite lot of resource versions since this cluster is quite old.

The database CPU usage is quite high (3 vCPUs, 4 GB RAM)
![image](https://user-images.githubusercontent.com/2479183/205685089-4528fac3-a5f1-4182-901d-52a8974deae3.png)

**Slow db logs**

```
2022-11-24 01:34:55 UTC [6129]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 LOG:  00000: process 6129 acquired ExclusiveLock on tuple (40,125) of relation 1733315 of database 1730368 after 1028.441 ms
2022-11-24 01:34:55 UTC [6129]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 LOCATION:  ProcSleep, proc.c:1495
2022-11-24 01:34:55 UTC [6129]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 STATEMENT:
            UPDATE resource_config_scopes
            SET last_check_start_time = now(), last_check_build_id = $1, last_check_build_plan = $2
            WHERE id = $3

2022-11-24 01:34:55 UTC [6150]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 LOG:  00000: process 6150 acquired ShareLock on transaction 2106307433 after 1119.776 ms
2022-11-24 01:34:55 UTC [6150]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 CONTEXT:  while updating tuple (241,88) in relation "resource_config_scopes"
2022-11-24 01:34:55 UTC [6150]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 LOCATION:  ProcSleep, proc.c:1495
2022-11-24 01:34:55 UTC [6150]: user=concourse,db=concourse,app=[unknown],client=51.159.74.196 STATEMENT:
            UPDATE resource_config_scopes
            SET last_check_start_time = now(), last_check_build_id = $1, last_check_build_plan = $2
            WHERE id = $3

```

**traces**

Some traces I managed to get with curl http://localhost:$CONCOURSE_DEBUG_BIND_PORT/debug/pprof/goroutine?debug=1  

```
goroutine profile: total 353596
349824 @ 0x43ca56 0x44d99e 0x44d975 0x46a585 0x478945 0x941e2e 0x941e0e 0x941cf7 0x941b26 0xc4e96d 0xe966ef 0xe8a89e 0x46e781
#   0x46a584    sync.runtime_SemacquireMutex+0x24                           runtime/sema.go:77
#   0x478944    sync.(*Mutex).lockSlow+0x164                                sync/mutex.go:171
#   0x941e2d    sync.(*Mutex).Lock+0x6d                                 sync/mutex.go:90
#   0x941e0d    github.com/concourse/concourse/atc/db/lock.(*lock).Acquire+0x4d             github.com/concourse/concourse/atc/db/lock/lock.go:217
#   0x941cf6    github.com/concourse/concourse/atc/db/lock.(*lockFactory).Acquire+0x156         github.com/concourse/concourse/atc/db/lock/lock.go:175
#   0x941b25    github.com/concourse/concourse/atc/db/lock.lockFactories.Acquire+0xa5           github.com/concourse/concourse/atc/db/lock/lock.go:161
#   0xc4e96c    github.com/concourse/concourse/atc/db.(*inMemoryCheckBuild).AcquireTrackingLock+0xec    github.com/concourse/concourse/atc/db/build_in_memory_check.go:406
#   0xe966ee    github.com/concourse/concourse/atc/engine.(*engineBuild).Run+0x16e          github.com/concourse/concourse/atc/engine/engine.go:115
#   0xe8a89d    github.com/concourse/concourse/atc/builds.(*Tracker).trackBuild.func1+0x31d     github.com/concourse/concourse/atc/builds/tracker.go:109

820 @ 0x43ca56 0x44c8dc 0xe96de5 0xe8a89e 0x46e781
#   0xe96de4    github.com/concourse/concourse/atc/engine.(*engineBuild).Run+0x864      github.com/concourse/concourse/atc/engine/engine.go:218
#   0xe8a89d    github.com/concourse/concourse/atc/builds.(*Tracker).trackBuild.func1+0x31d github.com/concourse/concourse/atc/builds/tracker.go:109
```

![image](https://user-images.githubusercontent.com/2479183/205686062-b1aaa8dc-512c-4f90-9df2-50b0d0167bad.png)

**Workaround**

We managed to find a workaround by increasing component-runner-interval from 10s to 30s using 
 `CONCOURSE_COMPONENT_RUNNER_INTERVAL : 30s`

The topic has been raised on Discord here https://discord.com/channels/219899946617274369/413770960089382922/1045614009317077022

## Triaging info

* Concourse version: 7.8.3
* Browser (if applicable):
* Did this used to work?

**Comment by maksymiliev** [comment]
Having the same issue on v7.9.1
```
 concourse-web.service - Concourse CI Web
     Loaded: loaded (/etc/systemd/system/concourse-web.service; enabled; vendor preset: enabled)
     Active: failed (Result: oom-kill) since Fri 2023-06-23 19:51:33 UTC; 2 days ago
    Process: 816 ExecStart=/usr/local/concourse/bin/concourse-web (code=killed, signal=TERM)
   Main PID: 816 (code=killed, signal=TERM)
        CPU: 9h 14min 16.616s

Jun 23 19:51:26 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:25.825590824Z","level":"info","source":"atc","message":"atc.tracker-imb.run.running","data":{"build":"check","pipeline":"rc-api>
Jun 23 19:51:26 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:26.440481628Z","level":"info","source":"tsa","message":"tsa.connection.channel.command.heartbeat.start","data":{"command":"forw>
Jun 23 19:51:27 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:27.534104979Z","level":"info","source":"atc","message":"atc.tracker-imb.run.running","data":{"build":"check","pipeline":"autota>
Jun 23 19:51:28 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:28.274958502Z","level":"info","source":"atc","message":"atc.tracker-imb.run.check-step.initializing","data":{"build":"check","p>
Jun 23 19:51:29 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:29.107718765Z","level":"info","source":"atc","message":"atc.tracker-imb.run.check-step.initializing","data":{"build":"check","p>
Jun 23 19:51:30 concourse-web concourse-web[842]: {"timestamp":"2023-06-23T19:51:29.555212197Z","level":"info","source":"atc","message":"atc.tracker-imb.run.check-step.initializing","data":{"build":"check","p>
Jun 23 19:51:31 concourse-web systemd[1]: concourse-web.service: A process of this unit has been killed by the OOM killer.
Jun 23 19:51:32 concourse-web systemd[1]: concourse-web.service: Killing process 1199 (concourse) with signal SIGKILL.
Jun 23 19:51:33 concourse-web systemd[1]: concourse-web.service: Failed with result 'oom-kill'.
Jun 23 19:51:33 concourse-web systemd[1]: concourse-web.service: Consumed 9h 14min 16.616s CPU time.
```

**Comment by taylorsilva** [comment]
Based on the traces and pprof graph you collected, it seems like the issue is that there are too many go routines trying to acquire the same lock. The build tracker ComponentRunner gets run so frequently that there isn't enough time for all the go routines to acquire the lock and do their work. I'd also guess that there are a large number of checks always being run.

So you have this constant drip of work coming in from the in-memory checks channel:
https://github.com/concourse/concourse/blob/55ff4b2410cb67746c14438a0af658151deb20b0/atc/builds/tracker.go#L125-L130

And then every 10s you kick-off one go routine for each build in the "starting" state:
https://github.com/concourse/concourse/blob/55ff4b2410cb67746c14438a0af658151deb20b0/atc/builds/tracker.go#L60-L70

Each call to `bt.trackBuild()` also doesn't wait for that work to finish. It gets started and then the component runner says "I'm done!" when it actually isn't done. If the previous call to `Run()`'s go routines aren't done we could end up kicking off more go routines than we need to. If all go routines from the previous call to `Run()` take longer than 10s to finish I think it would cause an increasing amount of go routines to start and get stuck.

One way I think we could solve this is by waiting for the current set of go routines to all finish before returning from `Run()`. This may mean on a very busy cluster that it may take longer than 10s for a build to actually start. That's probably more desirable than having your web node OOM killed though.

**Comment by taylorsilva** [comment]
I've opened a PR that I'm _hopeful_ will resolve this.

I want to be clear that I could **not** reproduce this issue. I have a feeling I would need a Concourse with A LOT of resource checks always running.

Based on the info OP provided I developed a theory for what may be happening (see PR for details). I was able to write a test that validated this theory. Is this the cause of the ever increasing amounts of go routines? I don't know. Once the PR is merged this issue will be closed.

If people are still seeing this issue after v7.13.0 is released then feel free to re-open or create a new issue.
